### ISSUE TITLE
What are the (so) links for?

### ISSUE THREAD

**Issue by paulo-ferraz-oliveira** [JeromeDeBretagne/erlanglauncher#5]
Hi.

First time using your project. Worked flawlessly (got the pong), and I was able to tweak it to compile `hello_jinterface.erl` at creation time and run according to that. Great!

Now I'm looking closer at the code and, given my reduced knowledge of Android internals, I'm wondering why 

```java
File filesMappingFile = new File(info.nativeLibraryDir, "libmappings.so");
...
```

is required. I can't find information on `libmappings.so` or why does `←` work (since I don't know the file format).

Also, it seems we're creating these "links" every time we boot the app. Is this actually necessary?

Thanks.

**Comment by paulo-ferraz-oliveira** [comment]
I found the `mappings.so` internally, in `build.gradle`, so also its format. That's already something.

**Comment by paulo-ferraz-oliveira** [comment]
I'm still trying to understand what all the links for all the `so`s are for, though.

**Comment by JeromeDeBretagne** [comment]
Hi @paulo-ferraz-oliveira ,

I believe you should find most/all the details in this previous issue #2 and in particular in [this comment](https://github.com/JeromeDeBretagne/erlanglauncher/issues/2#issuecomment-725441170). Let me know if you have further questions after going through this previous conversation.

> Also, it seems we're creating these "links" every time we boot the app.
 
Indeed, that's the behavior of the current code.

> Is this actually necessary?

No, not necessary at all, this was a design limitation of the original sample code when I forked it. The current approach was kept for its simplicity. There is even a comment specifically about this limitation in the code here :
https://github.com/JeromeDeBretagne/erlanglauncher/blob/26e42654925474fa5c3911d69cdcb601ccfb639c/app/src/main/java/com/isotope11/erlanglauncher/MainActivity.java#L93-L94

This would clearly deserve an improvement in production code, to have a trigger to decide when to create/update the links when the Erlang runtime packages are updated, either simply based on versioning or with some more-advanced logic. 

Pull requests are welcome :smiley: especially if we want to get more people to use Erlang/OTP on Android devices!

**Comment by paulo-ferraz-oliveira** [comment]
Great. Thanks for the comments (I did look at the Closed Issues' titles this afternoon, but didn't open them 😢). I'm fully enlightened on this subject, now.

> // Need to make this call optional, check if it's there, or something... 

I was thinking this could be supported by a version-specific check: let's say `erlanglauncher` is versioned (automatically) in an asset that is copied over to the `data` folder. If, when launching again, the same version is detected nothing happens; otherwise the whole system is redone.

I'm not using it for production code, though, so it's not an immediate issue. I'm fiddling around with it for an hackathon (I jokingly said I wanted to run Erlang in a mobile device and ended up doing it 😄).

For future reference, this is what I think would be nice to have:
1. copying of the `.so` files only when required,
2. a bundled dependency management system (e.g. `rebar3`),
3. this lib. made more generic in order to be used as a base for building stuff (rather than just run a simple ping-pong example*),
4. make this work in the emulator (I can make it work in a device, but somehow it fails to execute stuff on the emulator),
5. describe the Erlang compile+zip process

---

*which, don't get me wrong, is great, for starters and initial motivation.

**Comment by JeromeDeBretagne** [comment]
Thanks for all the suggestions !

**Comment by paulo-ferraz-oliveira** [comment]
> Thanks for all the suggestions !

Sure :-) It's also a guide for future me, since I'll try to have `rebar3` working on Android tomorrow :-)

**Comment by paulo-ferraz-oliveira** [comment]
We spent _some time_ getting to have a fully working x86_64 Android version on the emulator (we could only work on a real device before, due to the x86/ARM disparity), and were eventually able to - but this left no time to explore having `rebar3` inside the "device/emulator". 😢 

Changes to `MainActivity.java`/`build.gradle` were minor (new ERTS for 23.2.1 and new zip+arch location/naming for the compiled sources).

We had to create an `xcomp.conf` file for the build-from-source procedure* and change stuff to accept `ERL_ROOTDIR` for `erl` and `start` (vide https://github.com/erlang/otp/pull/2863).

```erlang
if [ -z "$ERL_ROOTDIR" ]
then
    ROOTDIR="/data/data/your.package.name/files/erlang"
else
    ROOTDIR="$ERL_ROOTDIR"
fi
```

(^ this is what we did, as per your example).

Note that our `xcomp.conf` file refers to `NDK_ROOT`, not `NDK_SYSROOT` as the [OTP doc. implies](https://github.com/erlang/otp/blob/master/HOWTO/INSTALL-ANDROID.md#define-system-variables).

___

* I'll try to pull request this into Erlang/OTP: don't know if it's something other people are interested in.

**Comment by paulo-ferraz-oliveira** [comment]
Here's the procedure we followed (kind of like a guide for us):

```
Ref: https://github.com/erlang/otp/blob/master/HOWTO/INSTALL-ANDROID.md

export NDK_ROOT=~/Library/Android/sdk/ndk/22.0.7026061
export PATH=$NDK_ROOT/toolchains/llvm/prebuilt/darwin-x86_64/bin:$PATH
export NDK_ABI_PLAT=android29

./otp_build autoconf # you need otp from Git @ erlang/otp and choose an appropriate tag
./otp_build configure --with-ssl=/usr/local/opt/openssl@1.1 --disable-dynamic-ssl-lib --disable-hipe --without-termcap --without-wx --enable-builtin-zlib --with-javac=JAVAC --xcomp-conf=./xcomp/erl-xcomp-x86_64-android.conf

export ERL_TOP=`pwd`

make
make release
./Install -cross -minimal .
```

(it would be interesting to have this documented somewhere, in `erlanglauncher` so it's easy to replicate the steps)

**Comment by paulo-ferraz-oliveira** [comment]
We accomplished this, from the previous action list:
* describe the Erlang compile+zip process,
* make this work in the emulator (I can make it work in a device, but somehow it fails to execute stuff on the emulator).

Missing:
* copying of the `.so` files only when required,
* a bundled dependency management system (e.g. `rebar3`),
* this lib. made more generic in order to be used as a base for building stuff (rather than just run a simple ping-pong example).

**Comment by JeromeDeBretagne** [comment]
Thanks a lot @paulo-ferraz-oliveira for the detailed feedback, that's super intereresting. I hadn't planned to include an x86_64 version as there is almost no x86_64 Android mobile devices left anymore, but I hadn't planned the emulator usage, that's a very valid case indeed!

> We had to create an `xcomp.conf` file for the build-from-source procedure* and change stuff to accept `ERL_ROOTDIR` for `erl` and `start` (vide [erlang/otp#2863](https://github.com/erlang/otp/pull/2863)).

Quick question on this one, why did you compile a cross build and didn't use a native version? Is it because you were on a Mac computer? As my host is an x86_64 Linux computer, I would have spontaneously re-used a native build and expect it to work on an x86_64 Android device, but that's maybe a shortcut.   

> Note that our `xcomp.conf` file refers to `NDK_ROOT`, not `NDK_SYSROOT` as the [OTP doc. implies](https://github.com/erlang/otp/blob/master/HOWTO/INSTALL-ANDROID.md#define-system-variables).

I'm confused about this comment, you should have both an `NDK_ROOT` and and `NDK_SYSROOT` in your `xcomp.conf` like on this line: https://github.com/erlang/otp/blob/fccb8482efc47bf2e48911564df502f087cce5ed/xcomp/erl-xcomp-arm64-android.conf#L77
as there should be a directory named `sysroot` at the root of your NDK path. Isn't it the case?

> (it would be interesting to have this documented somewhere, in erlanglauncher so it's easy to replicate the steps)

This looks better suited as a documentation in the Erlang/OTP project, as erlanglauncher is trying to focus on the specificities to *run* Erlang/OTP on Android (but not on how to build Erlang/OTP for Android).

Cheers, Jérôme

**Comment by paulo-ferraz-oliveira** [comment]
> Thanks a lot @paulo-ferraz-oliveira for the detailed feedback, that's super intereresting. I hadn't planned to include an x86_64 version as there is almost no x86_64 Android mobile devices left anymore, but I hadn't planned the emulator usage, that's a very valid case indeed!

There are "almost no x86_64 Android mobile devices left", I know, but:
1. it was also an exercise, to learn how to do it,
2. for some reason, the ARM emulator wasn't working initially, and (after some tweaking), when it did start working, it was super-slow.

The emulator usage is quite nice, yes, especially in development teams when the devices are limited.

> Quick question on this one, why did you compile a cross build and didn't use a native version? Is it because you were on a Mac computer?

Yes.

> As my host is an x86_64 Linux computer, I would have spontaneously re-used a native build and expect it to work on an x86_64 Android device, but that's maybe a shortcut.

I though about not using my work (Mac) computer for this, and my personal (Linux) one, but, as you say, it's kind of shortcut-ish. Also, there's dynamic linking to take into account, if we're using it (I don't think we did, and I don't think I do, locally, either - but it might be an issue), and file paths that may need to be tweaked, as you pull requested already.

> I'm confused about this comment, you should have both an NDK_ROOT and and NDK_SYSROOT in your xcomp.conf like on this line: https://github.com/erlang/otp/blob/fccb8482efc47bf2e48911564df502f087cce5ed/xcomp/erl-xcomp-arm64-android.conf#L77
as there should be a directory named sysroot at the root of your NDK path. Isn't it the case?

Well, I installed `ndk` from Android Studio, so couldn't quite figure where the root was, but yes, `sysroot` exists and the `xcomp.conf` file I submitted for OTP review makes use of it (I later downloaded `ndk` directly from the web the confirm my suspicion about that folder). Android Studio seems to tailer the `ndk` to whatever platform is pulling it from the web (in my case (`prebuilt/`)`darwin-x86_64`.

> This looks better suited as a documentation in the Erlang/OTP project, as erlanglauncher is trying to focus on the specificities to run Erlang/OTP on Android (but not on how to build Erlang/OTP for Android).

Right you are, and, apart from some minor details, it does already. (I'm not sure `ERL_TOP` is required, even).

**Comment by JeromeDeBretagne** [comment]
For future reference, the xcomp.conf mentioned above to cross compile for x86_64 Android was proposed in this pull request: https://github.com/erlang/otp/pull/2963 and was merged in the master branch, adding the following commit: https://github.com/erlang/otp/commit/da31dbea71ff2f50839e96bc6e6d11327bb868f8 . With a huge delay, many thanks @paulo-ferraz-oliveira !

**Comment by D4no0** [comment]
I tried to follow the process to compile OTP described here and I cannot get it working, I have the following setup:

OTP: 24.0
NDK: 22.1.7171670

The logs are the following:
```

 CXX	obj/x86_64-pc-linux-android/opt/jit/asmjit/core/virtmem.o
asmjit/core/virtmem.cpp:262:14: error: use of undeclared identifier 'shm_unlink'; did you mean 'sem_unlink'?
      return shm_unlink(tmpName.data());
             ^~~~~~~~~~
             sem_unlink
/home/daniel/Android/Sdk/ndk/22.1.7171670/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/include/semaphore.h:77:5: note: 'sem_unlink' declared here
int sem_unlink(const char* __name);
    ^
asmjit/core/virtmem.cpp:453:21: error: use of undeclared identifier 'shm_open'
      anonMem->fd = shm_open(anonMem->tmpName.data(), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
                    ^
2 errors generated.
```

**Comment by JeromeDeBretagne** [comment]
Hi @D4no0,

Good catch. These 2 functions are not supported by the Android NDK, this is officially documented here:
https://android.googlesource.com/platform/bionic/+/master/docs/status.md
```
Missing functions are [...] explicitly disallowed by SELinux:
    shm_open/shm_unlink
```
and they are commented out in the code here:
https://android.googlesource.com/platform/bionic/+/master/tests/headers/posix/sys_mman_h.c#90
```
#if !defined(__BIONIC__) // Disallowed by SELinux, so not implemented.
  FUNCTION(shm_open, int (*f)(const char*, int, mode_t));
  FUNCTION(shm_unlink, int (*f)(const char*));
#endif
```

I guess there are a few options for OTP 24, the quick fix currently should be to disable JIT when cross-compiling for Android x86_64. A better one in the long-term would be to see if there are ways to avoid using these 2 functions conditionnally, as at some point JIT will be enabled for 64-bit ARM processors too, and getting the JIT performance boost would be really nice.

**Comment by D4no0** [comment]
As far as I read about the `shm_open`, it is considered a bad practice and will never be added to Android. Maybe there is a possibility to implement this without using shared memory?

**Comment by JeromeDeBretagne** [comment]
If you look at the implementation, there is already some conditional code using `#if defined` pre-processor lines to handle the different implementations for BSD, MacOS, Linux, etc. :
https://github.com/erlang/otp/blob/master/erts/emulator/asmjit/core/virtmem.h
https://github.com/erlang/otp/blob/master/erts/emulator/asmjit/core/virtmem.cpp

In fact, there are already some code parts that don't use shm_open, for instance here:
https://github.com/erlang/otp/blob/master/erts/emulator/asmjit/core/virtmem.cpp#L442-L449
```
    if (!ASMJIT_VM_SHM_DETECT || preferTmpOverDevShm) {
      anonMem->tmpName.assign(VirtMem_getTmpDir());
      anonMem->tmpName.appendFormat(kShmFormat, (unsigned long long)bits);
      anonMem->fd = open(anonMem->tmpName.data(), O_RDWR | O_CREAT | O_EXCL, 0);
      if (ASMJIT_LIKELY(anonMem->fd >= 0)) {
        anonMem->fileType = AnonymousMemory::kFileTypeTmp;
        return kErrorOk;
      }
    }
```

however the following `else` block still makes a call to `shm_open` which is not defined on Android 
```
    else {
      anonMem->tmpName.assignFormat(kShmFormat, (unsigned long long)bits);
      anonMem->fd = shm_open(anonMem->tmpName.data(), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
      if (ASMJIT_LIKELY(anonMem->fd >= 0)) {
        anonMem->fileType = AnonymousMemory::kFileTypeShm;
        return kErrorOk;
      }
    }
```

so it still fails to compile due to the missing function.

We would have to first understand what's the best allocation strategy on Android to see how it should be implemented, if it can re-use an existing case or if it needs a dedicated one.

On first analysis, using a 'tmp' directory seems like an option to investigate: 

https://github.com/erlang/otp/blob/master/erts/emulator/asmjit/core/virtmem.h#L65-L72
```
  //! Not an access flag, only used by `allocDualMapping()` to override the
  //! default allocation strategy to always use a 'tmp' directory instead of
  //! "/dev/shm" (on POSIX platforms). Please note that this flag will be
  //! ignored if the operating system allows to allocate an executable memory
  //! by a different API than `open()` or `shm_open()`. For example on Linux
  //! `memfd_create()` is preferred and on BSDs `shm_open(SHM_ANON, ...)` is
  //! used if SHM_ANON is defined.
  kMappingPreferTmp = 0x80000000u
```

but clearly not by using `/tmp` since this directory doesn't exist on Android:
https://github.com/erlang/otp/blob/master/erts/emulator/asmjit/core/virtmem.cpp#L442-L449
```
#if !defined(SHM_ANON)
static const char* VirtMem_getTmpDir() noexcept {
  const char* tmpDir = getenv("TMPDIR");
  return tmpDir ? tmpDir : "/tmp";
}
#endif
```

There is maybe a better approach though. Then we will have to add additional preprocessor conditionals to remove the few remaining calls to undefined functions when compiling for Android. That sounds doable overall.

**Comment by D4no0** [comment]
I think for the moment JIT should be disabled for android, especially when there are no performance requirements on android emulator.

**Comment by JeromeDeBretagne** [comment]
You should be able to give it a try by passing the `--disable-jit` option to the configure step.

This is how it is documented at least : 
http://erlang.org/documentation/doc-12.0/erts-12.0/doc/html/BeamAsm.html

**Comment by D4no0** [comment]
Okay I can confirm that you can **build** the project when using `--disable-jit` command. Also I can confirm that erlang runtime **works** on an `x86_64` emulator after building it. The project was built without the `--with-ssl` flag.

I will make later a PR to update this example repository and one to update the documentation on how to build in otp repository.

**Comment by JeromeDeBretagne** [comment]
I've been able to cross-compile the Erlang runtime for x86_64 too, so I've updated the project with x86_64 support based on Erlang/OTP 24.0. This is implementing #8.

**Comment by JeromeDeBretagne** [comment]
> I will make later a PR to update this example repository and one to update the documentation on how to build in otp repository.

I've created 2 PRs upstream, erlang/otp#4832 to update the documentation and erlang/otp#4833 to disable the JIT compiler when cross-compiling for Android as it's not a platform supported by AsmJit at the moment.

**Comment by kobalicek** [comment]
@JeromeDeBretagne I saw the issue yesterday, no coincidence :)

I just fixed it so it would compile, but if I haven't tested it on Android myself. I think that the Linux/Android code path in general should go with `memfd_create`, which requires Linux 3.17+. The remaining code paths (`$TMPDIR` and `/dev/shm`) are fallbacks in case no other method works - if they fail then opening anonymous memory is not possible on that machine.

Anonymous memory is only needed for dual-mapping feature, which was written as a workaround for operating systems that enforce W^X hardening, but still allow to memory map the same region of memory twice. Well, I think somebody should play with it and see what works and what not. Dual mapping should really be seen as a last resort when it's not possible to allocate RWX memory with mmap().
